src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerConfigTests.cs

```diff
@@ -1471,5 +1471,318 @@ public void TreesShareOptionsInstances()
+                                (s) => Assert.Equal("/.globalconfig2", s));
+        }
+
+        [Fact]
+        public void DuplicateGlobalOptionsInNonGlobalConfigsAreKept()
+        {
+            var configs = ArrayBuilder<AnalyzerConfig>.GetInstance();
+            configs.Add(Parse(@"is_global = true
+option1 = value1", "/.globalconfig1"));
+
+            configs.Add(Parse(@"
+option1 = value2", "/.globalconfig2"));
+
+            var filteredSet = AnalyzerConfigSet.Create(configs.ToImmutableAndFree(), out var unsetKeys);
+            Assert.Empty(unsetKeys);
+        }
+
+        [Fact]
+        public void DuplicateSectionOptionsInNonGlobalConfigsAreKept()
+        {
+            var configs = ArrayBuilder<AnalyzerConfig>.GetInstance();
+            configs.Add(Parse(@"is_global = true
+[c:\path\to\file1.cs]
+option1 = value1
+", "/.globalconfig1"));
+
+            configs.Add(Parse(@"
+[c:\path\to\file1.cs]
+option1 = value2",
+"/.globalconfig2"));
+
+            var configSet = AnalyzerConfigSet.Create(configs.ToImmutableAndFree(), out var unsetKeys);
+            Assert.Empty(unsetKeys);
+        }
+
+        [Fact]
+        public void GlobalConfigsPropertiesAreGlobal()
+        {
+            var configs = ArrayBuilder<AnalyzerConfig>.GetInstance();
+            configs.Add(Parse(@"is_global = true
+option1 = value1
+", "/.globalconfig1"));
+
+            var options = GetAnalyzerConfigOptions(
+     new[] { "/file1.cs", "/path/to/file1.cs", "c:/path/to/file1.cs", "/file1.vb" },
+     configs);
+            configs.Free();
+
+            VerifyAnalyzerOptions(
+              new[]
+              {
+                    new[] { ("option1", "value1") },
+                    new[] { ("option1", "value1") },
+                    new[] { ("option1", "value1") },
+                    new[] { ("option1", "value1") }
+              },
+              options);
+        }
+
+        [Fact]
+        public void GlobalConfigsSectionsMustBeFullPath()
+        {
+            var configs = ArrayBuilder<AnalyzerConfig>.GetInstance();
+            configs.Add(Parse(@"is_global = true
+[c:/path/to/file1.cs]
+option1 = value1
+
+[*.cs]
+option2 = value2
+
+[.*/path/*.cs]
+option3 = value3
+
+[c:/.*/*.cs]
+option4 = value4
+", "/.globalconfig1"));
+
+            var options = GetAnalyzerConfigOptions(
+     new[] { "/file1.cs", "/path/to/file1.cs", "c:/path/to/file1.cs", "/file1.vb" },
+     configs);
+            configs.Free();
+
+            VerifyAnalyzerOptions(
+              new[]
+              {
+                    new (string, string)[] { },
+                    new (string, string)[] { },
+                    new (string, string)[]
+                    {
+                        ("option1", "value1")
+                    },
+                    new (string, string)[] { }
+              },
+              options);
+        }
+
+        [Fact]
+        public void GlobalConfigsSectionsAreOverriddenByNonGlobal()
```

Comment:
Do we also need tests that validate that ‘dotnet_diagnostic.DiagnosticId.severity’ entries in global analyzer configs are respected?

Reply:
They should be, because after being filtered they go through essentially the same mechanism. I'll add tests to verify though.

Reply:
Added tests to verify, which exposed an ordering problem + fixed it.